src/chips/sam4l/flashcalw.rs

```diff
@@ -0,0 +1,1103 @@
+                        self.lock_page_region(self.page.get(), true);
+                    },
+                    FlashState::Locking => {
+                        self.current_state.set(FlashState::Ready);
+                        self.current_command.set(Command::None);
+                    },
+                    _ => { assert!(false) /* should never reach here */ }
+
+                }
+            },
+            Command::Read => { 
+                // This isn't a real call and is handled synchronously (not here).
+                assert!(false);
+            },
+            Command::Erase => {
+                match self.current_state.get() {
+                    FlashState::Unlocking => {
+                        self.current_state.set(FlashState::Erasing);
+                        self.flashcalw_erase_page(self.page.get(), true); 
+                    }, 
+                    FlashState::Erasing => {
+                        self.current_state.set(FlashState::Locking);
+                        self.lock_page_region(self.page.get(), true);
+                    },
+                    FlashState::Locking => {
+                        self.current_state.set(FlashState::Ready);
+                        self.current_command.set(Command::None);
+                    },
+                    _ => { assert!(false); /* should never happen. */}
+                }
+            }
+            Command::None => { 
+                self.current_state.set(FlashState::Ready);
+            }
+
+        }
+        
+        //  If the command is finished call the complete CB.
+        if self.current_command.get() == Command::None && 
+            self.current_state.get() == FlashState::Ready {
+            self.client.map(|value| {
+                value.command_complete(Error::CommandComplete);
+            });
+        }
+    }
+  
+  
+    /// FLASH properties.
+    pub fn get_flash_size(&self) -> u32 {
+        let flash_sizes = [4,
+                           8,
+                           16,
+                           32,
+                           48,
+                           64,
+                           96,
+                           128,
+                           192,
+                           256,
+                           384,
+                           512,
+                           768,
+                           1024,
+                           2048];
+        // get the FSZ number and lookup in the table for the size.
+        flash_sizes[self.read_register(RegKey::PARAMETER) as usize & 0xf]  << 10
+    }
+
+    pub fn get_page_count(&self) -> u32 {
+        self.get_flash_size() / FLASH_PAGE_SIZE    
+    }
+
+    pub fn get_page_count_per_region(&self) -> u32 {
+        self.get_page_count() / FLASH_NB_OF_REGIONS
+    }
+
+
+    pub fn get_page_region(&self, page_number : i32) -> u32 {
+        (if page_number >= 0 
+            { unsafe { mem::transmute(page_number) } } 
```

Comment:
oh gosh, no! mem::transmute is super unsafe and should only be used when doing exceptional things like casting MMIO to a struct or something else not modeled in the language. This is just `page_number as u32`.